### PR TITLE
#184347019 implementing logout functionality

### PR DESCRIPTION
**Pull request by fabmukunzi**
### What does this PR do?
Create logout endpoint

### Description of Task to be completed?

- [x] Creates logout.controller.js file in controllers
- [x] Creates token.service.js file in services
- [x] Stores a jwt token in the jwtTokens relation as long as token is generated
- [x] Invalidates the token by changing token revoke status to true

### How should this be manually tested?

1. Clone [repo](https://github.com/atlp-rwanda/vikings-ec-bn.git)
2. Change into the directory **vikings-ec-bn**
3. Checkout to this branch: `feature/user-logout-184347019`
4. Run the command `npm install` to install packages
5. Run `npm run migrate`
6.Run `npm run dev`
7.In postman create a post request `localhost:5000/api/v1/users/logout`

### What are the relevant pivotal tracker stories?

[#184347019](https://www.pivotaltracker.com/story/show/184347019)

### CODE REVIEW

**Review by Fidela1 — APPROVED** [review]

**Review by Abdden — APPROVED** [review]

**Review by irakozetony — APPROVED** [review]

**Review by ndatumuremyi — APPROVED** [review]

File: src/middlewares/auth.middleware.js [hunk 1]
```diff
@@ -1,20 +1,28 @@
...
-    return res.status(201).json({ message: 'Unauthorized request' });
-
-	jwt.verify(token, process.env.SECRET_TOKEN, (err, user) => {
-		if (err)
-			return res.status(401).json({ message: 'Invalid or expired token'});
-		req.user = user;
-		next();
-	});
-};
+export const verifyToken = async(req, res, next) => {
```
Comment (ncutixavier) [inline src/middlewares/auth.middleware.js]: Rename this to `protectRoute`

File: src/controllers/logout.controller.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+            .then(() => {
+                res.status(200).json({
+                    message: "Logged out",
+                })
+            })
+
+    } catch (err) {
+        console.log(err)
+    }
+}
```
Comment (ncutixavier) [inline src/controllers/logout.controller.js]: - Follow the same convention applied in the user controller

File: src/middlewares/auth.middleware.js [hunk 3]
```diff
@@ -1,20 +1,27 @@
...
+}catch(err){
 
-	jwt.verify(token, process.env.SECRET_TOKEN, (err, user) => {
-		if (err)
-			return res.status(401).json({ message: 'Invalid or expired token'});
-		req.user = user;
-		next();
-	});
+    return res.status(500).json({ message:err?.message|| "something went wrong" });
+    }
```
Comment (ncutixavier) [inline src/middlewares/auth.middleware.js]: Let's enforce to have a proper error message

File: src/services/token.service.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { jwtTokens } from '../database/models/index.js';
+
+export class saveTokens {
+    static async saveToken(token) {
+        // let check = await jwtTokens.findOne({ where: { token: token } })
+        // if (check) {
+
+        // }
```
Comment (ncutixavier) [inline src/services/token.service.js]: Remove all unnecessary comments

File: src/utils/functions.util.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import {jwtTokens} from "../database/models/index"
+import { invalidateTokens } from "../services/token.service";
+export const checkRevoked=async(token)=>{
+    const tokenData=await jwtTokens.findOne({ where: { token: token } });
+    if(tokenData.dataValues.revoked===true){
+        return true;
+    }
+    return false;
+}
```
Comment (ncutixavier) [inline src/utils/functions.util.js]: - Don't use dataValues 'cause it's not necessary. Ask Yves for help
- Rename this file

**Review by ncutixavier — CHANGES_REQUESTED** [review]
All checks should be passed

**Review by ncutixavier — APPROVED** [review]

**Pull request merged**